nav2_dwb_controller/dwb_core/src/trajectory_utils.cpp

```diff
@@ -0,0 +1,107 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cmath>
+#include <dwb_core/trajectory_utils.hpp>
+#include <dwb_core/exceptions.hpp>
+#include <rclcpp/duration.hpp>
+
+namespace dwb_local_planner
+{
+const geometry_msgs::msg::Pose2D& getClosestPose(const dwb_msgs::msg::Trajectory2D& trajectory, const double time_offset)
+{
+  rclcpp::Duration goal_time = rclcpp::Duration::from_seconds(time_offset);
+  const unsigned int num_poses = trajectory.poses.size();
+  if (num_poses == 0)
+  {
+    throw nav2_core::PlannerException("Cannot call getClosestPose on empty trajectory.");
+  }
+  unsigned int closest_index = num_poses;
+  double closest_diff = 0.0;
+  for (unsigned int i=0; i < num_poses; ++i)
+  {
+    double diff = std::fabs((rclcpp::Duration(trajectory.time_offsets[i]) - goal_time).seconds());
+    if (closest_index == num_poses || diff < closest_diff)
+    {
+      closest_index = i;
+      closest_diff = diff;
+    }
+    if (goal_time < rclcpp::Duration(trajectory.time_offsets[i]))
+    {
+      break;
+    }
+  }
+  return trajectory.poses[closest_index];
+}
+
+geometry_msgs::msg::Pose2D projectPose(const dwb_msgs::msg::Trajectory2D& trajectory, const double time_offset)
+{
+  rclcpp::Duration goal_time = rclcpp::Duration::from_seconds(time_offset);
+  const unsigned int num_poses = trajectory.poses.size();
+  if (num_poses == 0)
+  {
+    throw nav2_core::PlannerException("Cannot call projectPose on empty trajectory.");
+  }
+  if (goal_time <= (trajectory.time_offsets[0]))
+  {
+    return trajectory.poses[0];
+  }
+  else if (goal_time >= rclcpp::Duration(trajectory.time_offsets[num_poses - 1]))
```

Comment:
Are these in nanoseconds? Make sure that you're using the right constructor. `Duration(number)` is nanoseconds, `Duration(numA, numB)` is seconds and then nanoseconds respectively.

Reply:
It is actually `Duration(bultin_interface/msg/duration)`

Reply:
Ah ok. I'd just ask you do a quick once over here and make sure all the nanoseconds <> seconds are considered.

Reply:
I think this is pretty legit in terms of considering nanoseconds. Here
```c++
Duration::Duration(
  const builtin_interfaces::msg::Duration & duration_msg)
{
  rcl_duration_.nanoseconds =
    static_cast<rcl_duration_value_t>(RCL_S_TO_NS(duration_msg.sec));
  rcl_duration_.nanoseconds += duration_msg.nanosec;
}
```

Reply:
I'm not saying for this instance, I'm just saying across your cherry pick. Since you ported this from ROS1 which uses seconds to ROS2 that uses nanoseconds, I'm just asking for a second look just to make sure.